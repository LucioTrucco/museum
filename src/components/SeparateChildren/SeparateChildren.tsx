import React, { useMemo, ReactNode, Children, ComponentType } from 'react';
import { ViewProps } from 'react-native';

type Props = {
  children?: ReactNode;
  Separator: ComponentType<ViewProps>;
};

export function SeparateChildren({ children, Separator }: Props) {
  const component = useMemo(() => {
    if (children === null || children === undefined) {
      return null;
    }

    const childrenArray = Children.toArray(children);
    return childrenArray
      .filter(Boolean)
      .reduce<Array<ReactNode>>(
        (r, n, idx) =>
          idx < childrenArray.length - 1
            ? [...r, n, <Separator key={`separator-${idx}`} />]
            : [...r, n],
        [],
      );
  }, [Separator, children]);

  return <>{component}</>;
}
