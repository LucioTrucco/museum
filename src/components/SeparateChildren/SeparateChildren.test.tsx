import React from 'react';
import { render } from '@testing-library/react-native';
import { SeparateChildren } from './SeparateChildren';
import { View } from 'react-native';

describe('tests => SeparateChildren', () => {
  // eslint-disable-next-line react-native/no-inline-styles
  const Separator = () => <View testID="separator" style={{ height: 10 }} />;

  it('should render SeparateChildren', () => {
    const component = render(
      <SeparateChildren Separator={() => <Separator />}>
        <View testID="contentChildren" />
        <View testID="contentChildren" />
        <View testID="contentChildren" />
      </SeparateChildren>,
    );
    expect(component.getAllByTestId('separator').length).toEqual(2);
    expect(component.getAllByTestId('contentChildren').length).toEqual(3);
  });

  it('should render with undefined children', () => {
    const component = render(
      // eslint-disable-next-line react/self-closing-comp
      <SeparateChildren Separator={() => <Separator />}></SeparateChildren>,
    );

    expect(component.container.props.children).toBe(undefined);
  });

  it('should render with null children', () => {
    const component = render(
      <SeparateChildren Separator={() => <Separator />}>
        {null}
      </SeparateChildren>,
    );

    expect(component.container.props.children).toBe(null);
  });

  it('should match snapshot', () => {
    const component = render(
      <SeparateChildren Separator={() => <Separator />}>
        {null}
      </SeparateChildren>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
