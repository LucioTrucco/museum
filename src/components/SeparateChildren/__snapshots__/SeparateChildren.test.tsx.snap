// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests => SeparateChildren should match snapshot 1`] = `null`;
